doc/api/esm.md

```diff
@@ -1154,6 +1154,40 @@ export async function transformSource(source,
 }
 ```
 
+#### <code>getGlobalPreloadCode</code> hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+Sometimes it can be necessary to run some code inside of the same global scope
+that the application will run in. This hook allows to return a string that will
+be ran as sloppy-mode script on startup.
+
+Similar to how CommonJS wrappers work, the last statement in the code may
+evaluate to a function. If it does, it will be called immediately. The only
+argument is a `require`-like function that can be used to load builtins.
+
+If the code needs more advanced `require` features, it will have to construct
+its own `require` using  `module.createRequire()`.
+
+```js
+/**
+ * @returns {string} Code to run before application startup
+ */
+export function getGlobalPreloadCode() {
+  return `\
+globalThis.someInjectedProperty = 42;
+console.log('I just set some globals!');
+
+(function (getBuiltin) {
+  const { createRequire } = getBuiltin('module');
+  const require = createRequire(process.cwd + '/<preload>');
+  // [...]
+});
```

Comment:
I think it would be simpler if we just treat `getBuiltin` as in scope for the global execution code.

Reply:
There's two ways of achieving that:

1. The preload code runs in an implicit function wrapper. It wouldn't be able to create `const`-style globals, at least not without some kind of eval.
2. The `getBuiltin` function is added temporarily to the global scope. I don't think that's something we'd want..?

Is that you arguing for the function wrapper..?

Reply:
It looks like you're suggesting finding a place where it's safe to do a contextual eval with only `getBuiltin` in scope. That feels pretty brittle to me but I may be wrong.

Reply:
Using `new Function('getBuiltin', code)` as @bmeck describes would achieve this.

Reply:
Right, so no const-style globals (option 1 above). I don’t think there’s a huge advantage of eval/Function vs CommonJS-style source wrapping..? The latter has the advantage of also working with that V8 flag that prevents string-based eval.

Reply:
Yes exactly - the main thing is to avoid the unnecessary need to write a function wrapper as above. The same sort of wrapper as CommonJS is fine too.

Reply:
I think I would just make it `(function (getBuiltin) { ${code}\n})`, reusing the same approach we use for CommonJS. Assumptions:

* We still want to call the function on the global scope (`fn.call(globalThis)`), not on some sort of custom object like with CommonJS.
* Since `getBuiltin` doesn't work like `require`, we don't want to call it `require`.
* We're fine with global `const` needing `vm.runInThisContext` from within the preload code. Implied: The preload code gets a function scope and top level code in general doesn't declare globals.

Reply:
@jkrems to my understanding we don't actually need to attach `runInThisContext` to any persisted global, is that correct? People can always use `this` or `globalThis` like you implied

Reply:
The only exception  is `const`/`let`-style globals which don't appear as properties on the global object. Those do require `runInThisContext` but that's a very narrow use case. I'd assume most preload code wouldn't want to declare them and is totally fine declaring `globalThis` properties.

Reply:
Not sure if `const`-style globals is the official name. What I mean is:

```console
$ node -e 'vm.runInThisContext("const x = 42;"); console.log(x, globalThis.x);'
42 undefined
$ node -e 'vm.runInThisContext("var x = 42;"); console.log(x, globalThis.x);'
42 42
```

Reply:
ah, function declarations are also a special kind of lexical binding. `let`/`const` are called variables in the global lexical scope in tc39 usually, but function declarations are different T_T (function declarations rules are stupidly complex in general)

Reply:
Done.